op-batcher/batch_submitter.go

```diff
@@ -253,58 +254,65 @@ mainLoop:
 				l.log.Error("issue getting adding L2 Block", "err", err)
 				continue
 			}
-			data := new(bytes.Buffer)
-			data.WriteByte(derive.DerivationVersion0)
-			if err := l.ch.OutputFrame(data, l.cfg.MaxL1TxSize); err != nil {
-				l.log.Error("error outputting frame", "err", err)
-				continue
-			}
+			// Hand role do-while loop to fully pull all frames out of the channel
+			for {
+				// Collect the output frame
+				data := new(bytes.Buffer)
+				data.WriteByte(derive.DerivationVersion0)
+				done := false
+				if err := l.ch.OutputFrame(data, l.cfg.MaxL1TxSize); err == io.EOF {
+					done = true
+				} else if err != nil {
+					l.log.Error("error outputting frame", "err", err)
+					continue mainLoop
+				}
 
-			// Poll for new L1 Block and make a decision about what to do with the data
+				// Query for the submitter's current nonce.
+				walletAddr := crypto.PubkeyToAddress(l.cfg.PrivKey.PublicKey)
+				ctx, cancel = context.WithTimeout(l.ctx, time.Second*10)
+				nonce, err := l.cfg.L1Client.NonceAt(ctx, walletAddr, nil)
+				cancel()
+				if err != nil {
+					l.log.Error("unable to get current nonce", "err", err)
+					continue
+				}
 
-			//
-			// Actually send the transaction
-			//
-			walletAddr := crypto.PubkeyToAddress(l.cfg.PrivKey.PublicKey)
+				// Create the transaction
+				ctx, cancel = context.WithTimeout(l.ctx, time.Second*10)
+				tx, err := l.CraftTx(ctx, data.Bytes(), nonce)
+				cancel()
+				if err != nil {
+					l.log.Error("unable to craft tx", "err", err)
+					continue
+				}
 
-			// Query for the submitter's current nonce.
-			ctx, cancel = context.WithTimeout(l.ctx, time.Second*10)
-			nonce, err := l.cfg.L1Client.NonceAt(ctx, walletAddr, nil)
-			cancel()
-			if err != nil {
-				l.log.Error("unable to get current nonce", "err", err)
-				continue
-			}
+				// Construct the a closure that will update the txn with the current gas prices.
+				updateGasPrice := func(ctx context.Context) (*types.Transaction, error) {
+					l.log.Debug("updating batch tx gas price")
+					return l.UpdateGasPrice(ctx, tx)
+				}
 
-			ctx, cancel = context.WithTimeout(l.ctx, time.Second*10)
-			tx, err := l.CraftTx(ctx, data.Bytes(), nonce)
-			cancel()
-			if err != nil {
-				l.log.Error("unable to craft tx", "err", err)
-				continue
-			}
+				// Wait until one of our submitted transactions confirms. If no
+				// receipt is received it's likely our gas price was too low.
+				// TODO: does the tx manager nicely replace the tx?
+				//  (submit a new one, that's within the channel timeout, but higher fee than previously submitted tx? Or use a cheap cancel tx?)
+				ctx, cancel = context.WithTimeout(l.ctx, time.Second*time.Duration(l.cfg.ChannelTimeout))
+				receipt, err := l.txMgr.Send(ctx, updateGasPrice, l.cfg.L1Client.SendTransaction)
+				cancel()
+				if err != nil {
+					l.log.Error("unable to publish tx", "err", err)
```

Comment:
Once one tx goes missing, there's a missing frame, and it's not useful anymore to try and submit the next frames. It should continue to the `mainLoop` here.